### PR TITLE
Muon Background Corrections - Perform fit to find background

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR ensures the background of a fit is now calculated and populated into the data table. The background is not yet subtracted from any data - this will be done in a separate PR.

**To test:**
1. Comment out this line https://github.com/mantidproject/mantid/blob/4e3edf60bb556bdf1de046309270711e1ce1fc33/scripts/Muon/GUI/Common/corrections_tab_widget/background_corrections_view.py#L55
2. Open Muon Analysis and load PSI data
3. Go Corrections tab. Select `Auto` background corrections
4. Try the following combinations:
 - `Auto` with `Flat Background` between `5.0` and `9.8` should show `Correction skipped`
 - `Auto` with `Flat Background + Exp Decay` between `5.0` and `9.8` should show `Correction success`
 - `Auto` with `Flat Background` between `0` and `0.14` should show `Correction success` for timezero set to 0.0

5. Try unticking `Apply parameter changes to all domains`. Then change the start or end X of a row. A fit will be performed on the single row and update the `A0` in the table.

Part of #31816

*This does not require release notes* because **it is unfinished and still in development**.

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by robertapplin** [comment]
@AnthonyLim23 can you give this a quick test? In particular, the `A0 Error` is pretty much always zero so I don't know if it is useful info. Maybe we want the chi squared displayed instead because the corrections can be skipped if this is too large?

I increased the max chi squared from 10 in `PSIBackgroundSubtraction` to 50 in this because otherwise the correction would always be skipped

**Comment by AnthonyLim23** [comment]
Am I right in thinking this only fits a flat background?

**Comment by AnthonyLim23** [comment]
For just a flat background, you should be setting the timezero value to 0, then doing the correction from 0 to 0.15-ish. If you do the fit outside of the GUI you will get an answer with a reasonable error. However, doing the fit using the browser (I made a plot from the ADS then clicked fit) doesnt always seem to update with the correct parameters of errors.

**Comment by robertapplin** [comment]
It is ready for testing @AnthonyLim23 . I will add tests and update any failures this afternoon

File: scripts/Muon/GUI/Common/corrections_tab_widget/background_corrections_model.py [hunk 1]
```diff
@@ -107,20 +129,85 @@ def populate_background_corrections_data(self) -> None:
...
+            correction_data.setup_functions()
+            correction_data.status = f"Correction skipped - chi squared is poor ({chi_squared:.3f})."
+        elif "Failed to converge" in fit_status:
+            correction_data.setup_functions()
+            correction_data.status = f"Correction skipped - {fit_status}"
+        else:
+            if self._corrections_context.selected_function == FLAT_BACKGROUND:
+                correction_data.flat_background = function.clone()
+            elif self._corrections_context.selected_function == FLAT_BACKGROUND_AND_EXP_DECAY:
+                correction_data.flat_background = function.getFunction(0).clone()
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/corrections_tab_widget/background_corrections_model.py:176]: is it intended that we will always override the default functions after each fit?

File: scripts/Muon/GUI/Common/corrections_tab_widget/background_corrections_model.py [hunk 2]
```diff
@@ -107,20 +129,85 @@ def populate_background_corrections_data(self) -> None:
...
+        """Returns the fit function to use for a background fit."""
+        if self._corrections_context.selected_function == FLAT_BACKGROUND:
+            return correction_data.flat_background
+        elif self._corrections_context.selected_function == FLAT_BACKGROUND_AND_EXP_DECAY:
+            composite = FunctionFactory.createFunction("CompositeFunction")
+            composite.add(correction_data.flat_background)
+            composite.add(correction_data.exp_decay)
+            return composite
+
+        raise RuntimeError("The selected background function is not recognised.")
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/corrections_tab_widget/background_corrections_model.py:200]: can we have this in an else?

File: scripts/Muon/GUI/Common/corrections_tab_widget/background_corrections_view.py [hunk 3]
```diff
@@ -162,13 +191,28 @@ def end_x(self, run: str, group: str) -> float:
         """Returns the Start X associated with the provided Run and Group."""
         return float(self._table_item_value_for(run, group, END_X_COLUMN_INDEX))
 
+    def selected_start_x(self) -> float:
+        """Returns the Start X in the row that is selected."""
+        if self._selected_row is not None:
+            return float(self.correction_options_table.item(self._selected_row, START_X_COLUMN_INDEX).text())
+
+        raise RuntimeError("There is no selected run/group table row.")
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/corrections_tab_widget/background_corrections_view.py:199]: can we have an else?

File: scripts/Muon/GUI/Common/corrections_tab_widget/background_corrections_view.py [hunk 4]
```diff
@@ -162,13 +191,28 @@ def end_x(self, run: str, group: str) -> float:
...
+
+    def selected_end_x(self) -> float:
+        """Returns the End X in the row that is selected."""
+        if self._selected_row is not None:
+            return float(self.correction_options_table.item(self._selected_row, END_X_COLUMN_INDEX).text())
+
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/corrections_tab_widget/background_corrections_view.py:205]: can we have an else?

**Comment by AnthonyLim23** [comment]
If I unselect the apply to all domains and give each row a different start and end x.
Then I tick the apply to all domains
I then change the startX, then all of the end x's change too.
I think we should only update the start x's

If I load 2 runs with the apply to all ticked
Then update a start x value (With all runs unticked) - crash


The fit values look good

**Comment by robertapplin** [comment]
Thanks, its ready for testing again

**Review by AnthonyLim23 — DISMISSED** [review]

File: scripts/Muon/GUI/Common/thread_model_wrapper.py [hunk 5]
```diff
@@ -26,8 +26,12 @@ def execute(self):
 class ThreadModelWrapperWithOutput(object):
     def __init__(self, function=None):
         self.callback = function
+        self.args = None
         self.result = None
 
+    def set_args(self, *args):
```
Comment (martyngigg) [inline scripts/Muon/GUI/Common/thread_model_wrapper.py:32]: Can we simplify this and have everything done in 1 step in `__init__`? Args will be an empty tuple if nothing is passed?

```python
def __init__(self, function, *args):
        self.callback = function
        self.args = args
        self.result = None
```

It looks like the function would be better not to be optional here as it's not checked later?

For calling, an empty tuple will then expand to nothing with `*` so calling will be

```python
    def execute(self):
        self.result = self.callback(*self.args)
```

**Review by martyngigg — CHANGES_REQUESTED** [review]
I've just got 1 inline comment on something I think can be made more compact and simpler to understand.

File: scripts/Muon/GUI/Common/thread_model_wrapper.py (hunk 5 above)
Comment (robertapplin) [reply scripts/Muon/GUI/Common/thread_model_wrapper.py]: Thanks, that is much better!

**Review by martyngigg — APPROVED** [review]
Thanks for making these changes. Looks good now.

**Pull request merged**